kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/KafkaRequestDecoder.java

```diff
@@ -105,10 +119,24 @@ protected Frame decodeHeaderAndBody(ChannelHandlerContext ctx, ByteBuf in, final
             frame = opaqueFrame(in, correlationId, decodeResponse, length, hasResponse);
             in.readerIndex(sof + length);
         }
-
         return frame;
     }
 
+    // we make the smallest assumption we can that the message starts with apiKey + apiVersion + correlationId
+    private @NonNull DecodedRequestFrame<ApiVersionsRequestData> createV0ApiVersionRequestFrame(ChannelHandlerContext ctx,
+                                                                                                int correlationId) {
+        if (log().isTraceEnabled()) { // avoid boxing
+            log().trace("{}: downgrading apiVersion request to v0", ctx);
+        }
+        RequestHeaderData requestHeaderData = new RequestHeaderData();
+        requestHeaderData.setCorrelationId(correlationId);
+        requestHeaderData.setRequestApiKey(ApiKeys.API_VERSIONS.id);
+        requestHeaderData.setRequestApiVersion((short) 0);
+        ApiVersionsRequestData apiVersionsRequestData = SPECIAL_DOWNGRADED_REQUEST_DATA;
```

Comment:
You are using the singleton instance of SPECIAL_DOWNGRADED_REQUEST_DATA so that the DowngradeFilter can distinguish between a client using version 0 and the downgrade case.

An alternative would be to subclass ApiVersionsRequestData as DowngradedApiVersionsRequestData and let the  filter use an `instanceof`.

Reply:
Good idea Keith, I have reimplemented it this way and overridden a few methods so that if anything attempted to write the special message as byte the `size` and `write` methods will throw, as an extra defence in case we lost the short-circuiting in a refactor.